src/main/cpp/src/shuffle_split_detail.hpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/utilities/traits.hpp>
+
+namespace spark_rapids_jni {
+
+namespace detail {
+
+/**
+ * @brief Functor for computing size of data elements for a given cudf type.
+ *
+ * Note: columns types which themselves inherently have no data (strings, lists,
+ * structs) return 0.
+ */
+struct size_of_helper {
+  template <typename T>
+  constexpr std::enable_if_t<!cudf::is_fixed_width<T>() && !std::is_same_v<T, cudf::string_view>,
+                             size_t>
+  operator()() const
+  {
+    return 0;
+  }
+
+  template <typename T>
+  constexpr std::enable_if_t<!cudf::is_fixed_width<T>() && std::is_same_v<T, cudf::string_view>,
+                             size_t>
+  operator()() const
+  {
+    return sizeof(cudf::device_storage_type_t<int8_t>);
+  }
+
+  template <typename T>
+  constexpr std::enable_if_t<cudf::is_fixed_width<T>(), size_t> __device__
+  operator()() const noexcept
+  {
+    return sizeof(cudf::device_storage_type_t<T>);
+  }
+};
+
+/**
+ * @brief Header for each partition.
+ */
+struct partition_header {
+  uint32_t magic_number;
+  uint32_t version;
+  uint64_t offset;
+  uint64_t num_rows;
+  uint64_t validity_size;
+  uint64_t offset_size;
+  uint64_t data_size;
+};
+
+// padding values for each validity type, as applied at the end of that data type
+// in each partition. so for example all of the grouped-together validity buffers for
+// a given partition will have a final 4 byte pad applied before the offset buffers begin
+constexpr size_t validity_pad = 4;
+constexpr size_t offset_pad   = 8;
```

Comment:
The kudo serializer indicates that this and the data padding should be 4 bytes, not 8. Is there a reason for 8, like the decimal128?

https://github.com/NVIDIA/spark-rapids-jni/blob/74e33d948442d72cd45dee2f3e062603d4953284/src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoSerializer.java#L129-L138

Reply:
Yeah it was just an optimal number I chose.  The original design doc didn't specify. I'll tweak them down - the code doesn't care: the offset copy only requires 4 bytes and the data copy uses the cub batched-memcpy which doesn't care at all. 

The above header sizes all came from the original design doc, so I guess that's a little behind. Easy tweaks.